File: backend/.NET/Calindra.Desafio/Calindra.Desafio.Domain/Results/Failure.cs
using Calindra.Desafio.Domain.Extensions;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;

namespace Calindra.Desafio.Domain.Results
{
    public class Failure
    {
        private readonly IList<string> _messages = new List<string>();

        protected Failure()
            => _messages = new List<string>();

        public Failure(params string[] messages)
            : this() => Add(string.Empty, messages);

        public Failure(string propertyName, params string[] messages)
            : this() => Add(propertyName, messages);

        public string Property { get; set; }

        public IReadOnlyCollection<string> Messages
            => new ReadOnlyCollection<string>(_messages);

        public void Add(string propertyName, params string[] messages)
        {
            Property = propertyName?.Trim();
            Add(messages);
        }

        public void Add(params string[] messages)
        {
            messages = messages?.Where(m => !m.IsEmpty()).ToArray();
            if (messages.IsEmpty())
                return;

            foreach (var message in messages)
                if (!_messages.Any(m => m.ToUpper().Equals(message, System.StringComparison.InvariantCultureIgnoreCase)))
                    _messages.Add(message);
        }
    }
}

File: backend/Node.JS/calindra-desafio/src/shared/results/failure.ts
export class Failure {
    private readonly _messages: string[] = []

    constructor(propertyName?: string) { 
        this.propertyName = propertyName?.trim() ?? '';
    }

    public propertyName: string;

    public get messages(): string[] {
        return new Array<string>(...this._messages);
    }

    public add(...messages: string[]) {
        const newMessages = messages.filter(message => !this._messages.includes(message));
        if (newMessages.length == 0)
            newMessages.push('Invalid value');
        newMessages.forEach(message => this._messages.push(message));
    }
}